### PR TITLE
[NativeAOT] Reflection type refactoring

### PR DESCRIPTION
**Pull request by jkotas**
Contributes to #91704

Design:

- `RuntimeType` - sealed light-weight System.Type, similar to CoreCLR RuntimeType. It has just two fields `MethodTable*` and lazily initialized pointer to the full reflection. The light-weight method and properties are implemented using `MethodTable*`, the rest initializes the full reflection `RuntimeTypeInfo` lazily and calls it to do the work.

- `RuntimeTypeInfo` - internal, lazily created full reflection. It is similar to the `Cache` attached to RuntimeType in CoreCLR. The reflection-free mode is implemented by blocking creation of `RuntimeTypeInfo` and throwing instead.

Some reflection micro-benchmark may regress with this change on native AOT. It is expected due to the extra indirections, but it makes the reflection more pay-for-play and a bit closer to how it is implemented in CoreCLR.

### CODE REVIEW

**Comment by jkotas** [comment]
I am not sure whether I like the `RuntimeType` and `MetadataOnlyType` split. It has been a bug farm, especially in the shared code. For example, there are number of checks like `type is RuntimeType` that tend to be broken when both `RuntimeType` and `MetadataOnlyType` represent `RuntimeType`.

**Comment by MichalStrehovsky** [comment]
> I am not sure whether I like the `RuntimeType` and `MetadataOnlyType` split. It has been a bug farm, especially in the shared code. For example, there are number of checks like `type is RuntimeType` that tend to be broken when both `RuntimeType` and `MetadataOnlyType` represent `RuntimeType`.

Hm, it was worth a shot. Maybe RuntimeType with a null type handle would still work. My concern with going that route was around another set of bugs (assuming we can get a type handle from a RuntimeType) but maybe there's fewer of those in the end.

> reflection-free mode is implemented by blocking creation

If reflection-free mode didn't survive this refactor I wouldn't be too sad :).

**Comment by EgorBo** [comment]
Does it mean we'll be able to enable some of the NonGC-related opts for RuntimeType now which currently only work on CoreCLR?

**Comment by jkotas** [comment]
> Does it mean we'll be able to enable some of the NonGC-related opts for RuntimeType now which currently only work on CoreCLR?

Yes, it is the goal for https://github.com/dotnet/runtime/issues/91704

**Comment by jkotas** [comment]
This should be ready to review.

Sanity check of binary sizes:
`dotnet new console --aot` want from 1,241,088 to 1,240,064.
`dotnet new console --aot` w/ `IlcDisableReflection=true` went from 1,005,056 to 994,816.

**Review by MichalStrehovsky — APPROVED** [review]
Looks good to the extent I can reasonably review!

File: src/coreclr/nativeaot/nativeaot.sln [hunk 1]
```diff
@@ -5,8 +5,6 @@ VisualStudioVersion = 16.0.30421.15
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "System.Private.CoreLib", "System.Private.CoreLib\src\System.Private.CoreLib.csproj", "{E4BC768B-F97D-4A8F-9391-B65DF3EB47C6}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "System.Private.DisabledReflection", "System.Private.DisabledReflection\src\System.Private.DisabledReflection.csproj", "{ADA691AE-4E1F-4212-97E6-51A27EFCE7E4}"
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/nativeaot/nativeaot.sln]: I have reverted this change and kept System.Private.DisabledReflection. Removing it broke things around Assembly and lazy P/Invoke too much for no good reason.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeType.cs [hunk 2]
```diff
@@ -1,16 +1,55 @@
...
 
 namespace System
 {
-    // Base class for runtime implemented Type
-    public abstract class RuntimeType : TypeInfo
+    internal sealed unsafe class RuntimeType : TypeInfo, ICloneable
     {
-        public sealed override string? GetEnumName(object value)
+        private MethodTable* _pUnderlyingEEType;
+        private RuntimeTypeInfo? _runtimeTypeInfo;
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeType.cs]: I made a quick change to use weak GC handle for the cached RuntimeTypeInfo. This new GC handle is redundant with the GC handle used by WeakReferences in the RuntimeTypeInfo unifiers. Something to keep in mind for future cleanups.

**Comment by jkotas** [comment]
The outerloop failures are unrelated. Fixing them in https://github.com/dotnet/runtime/pull/94213.

**Pull request merged**